for-upstream-only/nixos/profiles/cache-keys-json.nix

```diff
@@ -0,0 +1,138 @@
+# TODO (doc) CacheKeys format reference link
+/*
+  A module that configures an agent machine to use caches for reading and/or
+  writing, based on a secret JSON file in the CacheKeys format.
+
+  Like the format, it is currently limited to cachix.org caches.
+ */
+{ pkgs, lib, config, ...}:
+let
+  cfg = config.profile.cacheKeys;
+  inherit (lib.lists) filter concatMap concatLists;
+  inherit (lib) types isAttrs mkIf escapeShellArg attrValues mapAttrsToList;
+  inherit (builtins) readFile fromJSON map split;
+
+  readCacheKeys = fileOrNull:
+    if fileOrNull == null
+    then { kind = "CacheKeys"; caches = {}; }
+    else fromJSON (readFile fileOrNull);
+
+  json = readCacheKeys cfg.file;
+  inherit (json) caches;
+
+  pubkeys = concatMap (cache: cache.publicKeys) (attrValues json.caches);
+
+in
+{
+  options.profile.cacheKeys = {
+    file = lib.mkOption {
```

Comment:
Naming bikeshedding :) 

since the other option is called `deployedPath`, this one could be `evalPath`?

Reply:
I think the file vs. path distinction is good, even if not obvious, so that would make it `evalFile`.
It still doesn't seem a whole lot better, because we're not literally evaluating the file. I don't know what to do about this one.

Reply:
File is more precise, but then it should also be `deployedFile`?

Reply:
My reasoning is in the first case you use the actual file (including contents) whereas in the latter, you only provide a reference to those contents, the path.

Reply:
Normally the distinction is irrelevant in Nix, because (almost) everything can be treated as by-value. Paths and files are nearly equivalent when you automatically put contents in the Nix store.

Reply:
I think the user has to know two things:

a) what type it should be (string or path)
b) where the file should be located (machine to be evaluated on or the final target)

For (a) we can just assert and document, while (b) needs documentation but ideally would also be encoded in the name so it's harder to mess up.

Which file is actually read from is internal implementation detail and I don't think the user will really think that far (I didn't).

Reply:
`deployedPath` is an implementation detail when using the NixOps profile, so that's taken care of.
In other cases we don't know where the user will deploy it, so then the user will have to give that info.

The assertions and option documentation are in place. Frontend and docs site will be separate PRs, including one here for the docs site links.

User won't think about all the naming subtleties but just use the options, I agree. I don't think `evalFile` is sufficiently descriptive either, so I've accepted that the user will have to use documentation such as the options reference, or preferably an example from guide docs or frontend. `fileToReadDuringEvaluation` may be a bit too much. I don't think there's an optimum between this name and `file`, sadly.

Reply:
> deployedPath is an implementation detail when using the NixOps profile, so that's taken care of.

That's not the case for `darwin` and other agents. I'd like to reuse the same module to avoid duplicating code.

Maybe we should define state directory and provide default convention as a relative path to it?

Reply:
Concretely this may look like
 - `profile.nix` declares `profile.hercules-ci-agent.secretsDir` with a optionDefault value in /var/lib
 - `profile.nix` defines `services.hercules-ci-agent.*` based on secretsDir with normal priority (no `mkDefault` or similar)
 - `nixops-profile.nix` leaves secretsDir as is and reads its value indirectly from `services.hercules-ci-agent.*`. If user wants more indirection, `nixops-profile.nix` can not provide much value anyway.
 - nix-darwin imports `profile.nix`
 - nix-darwin sets secretsDir to ~ with `mkDefault` priority (overriding optionDefault)  - do we really want to keep doing it like [that](https://github.com/LnL7/nix-darwin/pull/141/files#diff-82ea436bf3c09a28badf08f02a970495R90) or in a subdir?

Furthermore, possibly off-topic:
 - do we need to split profile.nix into `profile.nix` and `darwin-profile.nix`, keeping the common only the common parts in `common.nix`. (I think we want to preserve that `profile.nix` is for NixOS)
 - on NixOS we have the user import these modules into their config, whereas with current nix-darwin, the module is upstreamed. Do we distribute everything through nix-darwin, or do we let the user pick a version and hope that it matches the upstreamed modules? How does it deal with its channel(s)?

Reply:
Rather than splitting up `profile.nix` into more pieces I'd prefer to merge it back into the module itself.

It can be a separate file what options darwin uses for easier configurability but NixOS option namespace wise it's best not to introduce two namespaces. 

For example, there's cognitive overhead thinking which namespace has which option, which just creates friction for the user.